[CMake] Formalize whether libdispatch is used/allowed

davezarzycki opened the pull request:
Accidentally requiring libdispatch is a repeated cmake regression. Let's avoid this entirely by formalizing whether libdispatch is enabled or not rather than infer it.

davezarzycki commented:
@swift-ci please test

buttaface commented on CMakeLists.txt:
```diff
@@ -442,14 +444,9 @@ if(SWIFT_ENABLE_EXPERIMENTAL_CONCURRENCY AND NOT SWIFT_STDLIB_SINGLE_THREADED_RU
-if(SWIFT_BUILD_SYNTAXPARSERLIB OR SWIFT_BUILD_SOURCEKIT)
-  if(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
+if(SWIFT_ENABLE_DISPATCH AND NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
+  if(SWIFT_BUILD_SYNTAXPARSERLIB OR SWIFT_BUILD_SOURCEKIT OR SWIFT_CONCURRENCY_USES_DISPATCH)
```
I like the general idea of this pull, but this line subtly changes the logic: now if `SWIFT_CONCURRENCY_USES_DISPATCH` is true, then `SWIFT_BUILD_HOST_DISPATCH` is also set to true. But the two are independent, the former must only be set when the target concurrency libraries are built and the latter when the first two host libraries are built. This won't matter when you're only building everything for the host, but if you're cross-compiling the standalone stdlib, as I do for Android, this will pull in the host dispatch build unnecessarily.

davezarzycki committed:
[CMake] Formalize whether libdispatch is used/allowed

Accidentally requiring libdispatch is a repeated cmake regression. Let's
avoid this entirely by formalizing whether libdispatch is enabled or not
rather than infer it.

davezarzycki commented:
@buttaface – I think I've addressed your concerns. Feedback would be appreciated.

@swift-ci please test

finagolfin reviewed (approved):
Nice work simplifying the logic.

compnerd reviewed (changes requested):
This effectively disables the use of `SWIFT_BUILD_SYNTAXPARSERLIB` and `SWIFT_BUILD_SOURCEKIT` if `SWIFT_ENABLE_DISPATCH` is false.  I think that if you want to expose the explicit control over dispatch, please use `cmake_dependent_option` to ensure that the user cannot set `SWIFT_BUILD_SYNTAXPARSERLIB` or `SWIFT_BUILD_SOURCEKIT` while `SWIFT_ENABLE_DISPATCH` is a false value.

davezarzycki commented:
@compnerd – Thanks for the feedback. The patch now uses `cmake_dependent_option`.

@swift-ci please test

davezarzycki commented:
Hey @compnerd – It looks like that CMake feature is too new. From one of the Linux builders:
```
12:34:19 CMake Error at CMakeLists.txt:413 (cmake_dependent_option):
12:34:19   Unknown CMake command "cmake_dependent_option".
```

compnerd commented:
@davezarzycki I suspect you missed the `include(CMakeDependentOption)` which actually is required.  The feature has been there for a very long time (3.13 at the very least).

davezarzycki commented:
@compnerd – Ah that must be it. Fixed.

@swift-ci please test

compnerd commented:
@swift-ci please test Windows platform

compnerd reviewed (approved):
I think that this shouldn't change the default behaviour and still allow control over dispatch and validate inputs, so seems reasonable.  Please ensure that it passes the Windows CI as well (though I don't see why it wouldn't).
